### PR TITLE
Add control of Reset Policy to Adam optimizers

### PR DESCRIPTION
**Pull request by dtimson**
Seems to be needed to do "single-epoch loop" training (see mlpack issue #1455) without resetting each time.

[Duplicate of a PR in mlpack.]

### CODE REVIEW

**Comment by rcurtin** [comment]
You're right, we need to make sure `updatePolicy` gets initialized at least once.  So an `isInitialized` flag seems like the best solution to me.

I'll add your name during merge---thanks again!  Once this is merged I'll release ensmallen 1.12.0 with these changes.

Also, for mlpack first-time contributors we always offer to put some mlpack stickers in the mail that you could put on your laptop. But we don't have any ensmallen stickers... but if you want some mlpack stickers anyway I am happy to mail them to you; just send an email to ryan@ratml.org with your mailing address and I will get them sent out. :)

**Comment by rcurtin** [comment]
@dtimson: just wanted to check in, were you still planning to handle all the SGD variants here?  If not I can do it, just let me know.

**Comment by dtimson** [comment]
My apologies - I had got the impression you wanted to handle that yourself during merge. I'm happy to do it, leave it with me. To clarify, I will now: add ResetPolicy() to the optimizers; add resetPolicy to ctor args of the optimizers; add isInitialized flag to SGD; update optimzers.md doc file.

**Comment by zoq** [comment]
@dtimson Sounds like a good plan to me, looking forward to merge everything.

File: include/ensmallen_bits/ada_delta/ada_delta.hpp [hunk 1]
```diff
@@ -71,7 +71,8 @@ class AdaDelta
            const double epsilon = 1e-6,
            const size_t maxIterations = 100000,
            const double tolerance = 1e-5,
-           const bool shuffle = true);
+           const bool shuffle = true,
+           const bool resetPolicy = true);
```
Comment (zoq) [inline include/ensmallen_bits/ada_delta/ada_delta.hpp]: Can you add a description for  the ``resetPolicy``? That applies to the other optimizers as well.

File: include/ensmallen_bits/sgdr/snapshot_sgdr.hpp [hunk 2]
```diff
@@ -144,6 +145,11 @@ class SnapshotSGDR
     return optimizer.DecayPolicy().Snapshots();
   }
 
+  //! Get whether or not to accumulate the snapshots.
+  bool Accumulate() const { return accumulate; }
```
Comment (zoq) [inline include/ensmallen_bits/sgdr/snapshot_sgdr.hpp]: Nice catch 👍

File: include/ensmallen_bits/ada_delta/ada_delta.hpp (hunk 1 above)
Comment (dtimson) [reply include/ensmallen_bits/ada_delta/ada_delta.hpp]: Thanks. All done now.

**Comment by zoq** [comment]
Thanks again for the contribution 👍

**Pull request merged**